Room View

nreyes-qualabs opened the pull request:
first commit: rendering empty dummy divs

nreyes-qualabs committed:
WIP: rooms view add dummy participants

nreyes-qualabs committed:
better use of mui components + first attempt at adjusting size

nreyes-qualabs committed:
WIP: dynamic rendering strategy layed out w/ comments

nreyes-qualabs commented:
@mpollingerQualabs @ghbogarin 
I kinda laid out in comments the approach that I'll be trying to implement in this component. Would love if you guys can take a look and see if it makes sense and matches what we were talking about.

nreyes-qualabs committed:
properly arranging dummy tiles

nreyes-qualabs committed:
Merge branch 'develop' into 38-room-view

nreyes-qualabs committed:
replace text tiles with streamless videos

nreyes-qualabs committed:
using video components with user camera stream

nreyes-qualabs committed:
fix typo

marcovidonis commented on frontend/src/views/Room.jsx:
```diff
@@ -1,19 +1,107 @@
 
 function Room() {
+  const [idCount, setIdCount] = useState(0);
+  // const [participants, setParticipants] = useState([]);
```
The `Room` object from our `lib/webrtc` will give you a `remoteParticipants`, which is an array of `RemoteParticipant` objects that should include everything we need here.

marcovidonis commented on frontend/src/views/Room.jsx:
```diff
@@ -1,19 +1,107 @@
+  // if it ends up being a component property passed in by the parent component.
+  // for now it needs to be in the state so that the "Add Participant" test button can work
+  const [visibleParticipants, setVisibleParticipants] = useState([]);
+  const rowsLimit = { xs: 2, sm: 2, md: 2 };
```
I would propose a higher rows limit. This is basically a grid view now (we'll likely need something else for a presenter view etc), so we could do 3-4 rows easily. `xs` is a mobile device in portrait mode, so a 3x2 grid (3 rows, 2 columns) will probably work. On larger viewports, we could probably do 4 or 5 rows and perhaps reduce the max number of column to `{ xs: 2, sm: 4, md: 6 }` (`md` is essentially an iPad in portrait mode).

What do you all think?

nreyes-qualabs replied:
Sounds good to me, I had no idea what values to use so I randomly set it to {2,2,2} but it should work regardless

marcovidonis commented on frontend/src/views/Room.jsx:
```diff
@@ -1,19 +1,107 @@
+            </div>
+          </Grid>
+        ))}
+        <Grid item xs={1} sm={1} md={1}>
+          <div style={{ borderStyle: 'solid' }}>
+            <Video
+              stream={userStream}
+            />
+          </div>
+        </Grid>
```
Is this the tile for the 'self' video? Personally I'd prefer keeping it as an absolute-positioned element in the bottom right (for example) rather than as part of the tiles grid, but I'm interested in hearing what people think.

Also, let's consider we'll have different view layouts (grid and presenter at least, I believe), so we might just adopt different strategies in each.

mpollingerQualabs replied:
I agree with having it bottom right, fixed position always

nreyes-qualabs replied:
Yeap I also agree. By placing it at the end of the gird, because of how the grid gets rendered, it will always get placed at the bottom right of the grid. Is that sufficient for now or do we need to take it out of the grid for some other reason?

nreyes-qualabs replied:
The only way it wouldn't end up in the bottom right is if the precondition of `visibleParticipants` <= `rowsLimit*tilesPerRowLimit` fails to be met

nreyes-qualabs replied:
Nevermind I think I've got what you guys are saying, completely out of the grid, fixed in the bottom right of the view. Got it

marcovidonis replied:
Right. What I mean by fixed position is out of the grid, but that's a matter of personal preference. We could always get back to it later if we want :)

Personally the downsides I see with having your own video as part of the grid are:
- we "lose" one slot for a potential remote stream that we need to hide, so that the own video is always visible
- the position of the local video in the page will move as people join and leave, so you sort of have to look for it if you want to see yourself (minor issue, but adds a little friction)
- you don't get a clear separation of local streams vs remote streams. Potentially there are cases in which you might be sharing more than one stream (video and screenshare, for example), so a clearer separation helps

But as I said, these are not major issues, so we can just as well keep the grid as it is, if it means we can move on faster with the project ;)

nreyes-qualabs replied:
Yes I see now, I agree we should take it out of the grid for sure

marcovidonis reviewed (commented):
I think it's looking good! Feel like plugging in the WebRTC part now? The RoomTest component gets the JWT from a .env file, and I think we can set env vars in Netlify for the build. We could use the JWT I used in my tests for now, until we can get it from edge functions.

nreyes-qualabs committed:
wip, trying to connect to mux room

nreyes-qualabs committed:
add leaveRoom() for unmount

nreyes-qualabs committed:
add subscription to all remote participants

nreyes-qualabs commented:
succesfully connected to mux room using JWT set with env variable

nreyes-qualabs committed:
make video component work with % sizing

nreyes-qualabs committed:
WIP testing room view

nreyes-qualabs committed:
change computeTilesPerRow approach, working the best so far

nreyes-qualabs committed:
remove testing code

nreyes-qualabs committed:
Merge branch 'develop' into 38-room-view

nreyes-qualabs committed:
get JWT with post request using roomId from path

nreyes-qualabs committed:
add return value to get jwt action

nreyes-qualabs committed:
room view properly connecting to given room
+ redirects to roomNotFound view when get jwt method returns 404

nreyes-qualabs committed:
generateToken returns more accurate http error codes

nreyes-qualabs committed:
Merge branch 'develop' into 38-room-view

# Conflicts:
#	frontend/package-lock.json
#	frontend/src/actions/index.js
#	frontend/src/index.jsx

nreyes-qualabs committed:
fix mistaken deletion of createRoom and addRoomToDb methods

marcovidonis commented on frontend/src/views/Room.jsx:
```diff
@@ -1,19 +1,168 @@
+              }
+            </Grid>
+            <div style={localStreamStyle}>
+              <Video
+                stream={localStream}
+              />
```
This will prevent the local user from hearing themselves:

```suggestion
              <Video
                 stream={localStream}
                 muted
               />
```

marcovidonis commented:
There are 3 problems currently:

- [x] The local stream needs to be muted, or else the local user will hear themselves (see my code suggestion in the comment above)
- [x] We need a back button to return to `/rooms` that will also call `leaveRoom()`
- [x] ⚠️ There's an issue with subscribing with remote participants... If I join a room, I only seem to get the stream from one of them.

marcovidonis commented on frontend/src/views/Room.jsx:
```diff
@@ -1,19 +1,168 @@
+      const newParticipant = await newRoom.join();
+
+      newRoom.on('ParticipantTrackSubscribed', (remoteParticipant, track) => {
+        const stream = new MediaStream();
+        stream.addTrack(track.mediaStreamTrack);
+        const streamObj = { stream, participantId: remoteParticipant.id };
+        setRemoteStreams([...remoteStreams, streamObj]);
+      });
```
I'm not sure this is working... Also, here we create a stream for each track, rather than adding both audio and video tracks to the same stream, like we would with:

```js
tracks.forEach((track) => stream.addTrack(track.mediaStreamTrack));
```

marcovidonis commented on frontend/src/views/Room.jsx:
```diff
@@ -1,19 +1,168 @@
+      });
+
+      // newRoom.on('ParticipantJoined', (p) => console.log('someone joined', p));
+      // newRoom.on('ParticipantLeft', (p) => console.log('someone left', p));
```
We need to handle these events. We need to keep a state variable for remote participants, which we then update whenever a participant joins or leaves.
Also in these event handlers, subscribe/unsubscribe accordingly.

clararfe committed:
added mute flag

marcovidonis committed:
quick fix: subscribe/unsubscribe to remote participants

marcovidonis commented:
> There are 3 problems currently:
> 
> * [x]  The local stream needs to be muted, or else the local user will hear themselves (see my code suggestion in the comment above)
> * [x]  We need a back button to return to `/rooms` that will also call `leaveRoom()`
> * [x]  ⚠️ There's an issue with subscribing with remote participants... If I join a room, I only seem to get the stream from one of them.

I'll address the Leave button in the branch I'm working on where I'm adding the mute/unmute functions... I think it might end up looking something like the Discord UI:

<img width="550" alt="image" src="https://user-images.githubusercontent.com/31407403/219050961-8acae6c1-b105-4b65-b3e3-762b9d9db0c4.png">

marcovidonis committed:
fix leave room when component unmounts

nreyes-qualabs committed:
adding audio/video track to the same stream

marcovidonis committed:
force render update to show remote streams

marcovidonis committed:
actually fix showing remote streams

The previous hack failed when more than 2 participants joined,
or when a participant left and then joined again.
The new fix uses a reference to a map to keep track of the streams
from each remote participant, and then an array as a state variable
to display them.
